go/vt/vttablet/tabletmanager/vreplication/table_plan_partial.go

```diff
@@ -0,0 +1,209 @@
+
+		return false
+	}
+	return true
+}
+
+func (tpb *tablePlanBuilder) generatePartialValuesPart(buf *sqlparser.TrackedBuffer, bvf *bindvarFormatter, dataColumns *binlogdatapb.Bitmap) *sqlparser.ParsedQuery {
+	bvf.mode = bvAfter
+	separator := "("
+	for ind, cexpr := range tpb.colExprs {
+		if tpb.isColumnGenerated(cexpr.colName) {
+			continue
+		}
+		if !isBitSet(dataColumns.Cols, ind) {
+			continue
+		}
+		buf.Myprintf("%s", separator)
+		separator = ","
+		switch cexpr.operation {
+		case opExpr:
+			switch cexpr.colType {
+			case querypb.Type_JSON:
+				buf.Myprintf("%v", cexpr.expr)
+			case querypb.Type_DATETIME:
+				sourceTZ := tpb.source.SourceTimeZone
+				targetTZ := tpb.source.TargetTimeZone
+				if sourceTZ != "" && targetTZ != "" {
+					buf.Myprintf("convert_tz(%v, '%s', '%s')", cexpr.expr, sourceTZ, targetTZ)
+				} else {
+					buf.Myprintf("%v", cexpr.expr)
+				}
+			default:
+				buf.Myprintf("%v", cexpr.expr)
+			}
+		}
+	}
+	buf.Myprintf(")")
+	return buf.ParsedQuery()
+}
+
+func (tpb *tablePlanBuilder) generatePartialInsertPart(buf *sqlparser.TrackedBuffer, dataColumns *binlogdatapb.Bitmap) *sqlparser.ParsedQuery {
+	buf.Myprintf("insert into %v(", tpb.name)
+	separator := ""
+	for ind, cexpr := range tpb.colExprs {
+		if tpb.isColumnGenerated(cexpr.colName) {
+			continue
+		}
+		if !isBitSet(dataColumns.Cols, ind) {
+			continue
+		}
+		buf.Myprintf("%s%v", separator, cexpr.colName)
+		separator = ","
+	}
+	buf.Myprintf(")", tpb.name)
+	return buf.ParsedQuery()
+}
+
+func (tpb *tablePlanBuilder) generatePartialSelectPart(buf *sqlparser.TrackedBuffer, bvf *bindvarFormatter, dataColumns *binlogdatapb.Bitmap) *sqlparser.ParsedQuery {
+	bvf.mode = bvAfter
+	buf.WriteString(" select ")
+	separator := ""
+	for ind, cexpr := range tpb.colExprs {
+		if tpb.isColumnGenerated(cexpr.colName) {
+			continue
+		}
+		if !isBitSet(dataColumns.Cols, ind) {
+			continue
+		}
+		buf.Myprintf("%s", separator)
+		separator = ", "
+		buf.Myprintf("%v", cexpr.expr)
+
+	}
+	buf.WriteString(" from dual where ")
+	tpb.generatePKConstraint(buf, bvf)
+	return buf.ParsedQuery()
+}
+
+func (tpb *tablePlanBuilder) createPartialInsertQuery(dataColumns *binlogdatapb.Bitmap) *sqlparser.ParsedQuery {
+	bvf := &bindvarFormatter{}
+	buf := sqlparser.NewTrackedBuffer(bvf.formatter)
+
+	tpb.generatePartialInsertPart(buf, dataColumns)
+	if tpb.lastpk == nil {
+		// If there's no lastpk, generate straight values.
+		buf.Myprintf(" values ", tpb.name)
+		tpb.generatePartialValuesPart(buf, bvf, dataColumns)
+	} else {
+		// If there is a lastpk, generate values as a select from dual
+		// where the pks < lastpk
+		tpb.generatePartialSelectPart(buf, bvf, dataColumns)
+	}
+	return buf.ParsedQuery()
+}
+
+func (tpb *tablePlanBuilder) createPartialUpdateQuery(dataColumns *binlogdatapb.Bitmap) *sqlparser.ParsedQuery {
+	bvf := &bindvarFormatter{}
+	buf := sqlparser.NewTrackedBuffer(bvf.formatter)
+	buf.Myprintf("update %v set ", tpb.name)
```

Comment:
If it's not already, I think we should make it an IdentifierCS for the potential escaping.

Reply:
`colexpr` already uses `IndentifierCI`s